src/clp_ffi_py/ir/PyDecoderBuffer.cpp

```diff
@@ -0,0 +1,316 @@
+        m_read_buffer = new_buf;
+        m_buffer_capacity = new_capacity;
+    } else if (0 < num_unconsumed_bytes) {
+        memcpy(m_read_buffer, unconsumed_bytes, static_cast<size_t>(num_unconsumed_bytes));
+    }
+    m_num_current_bytes_consumed = 0;
+    m_buffer_size = num_unconsumed_bytes;
+
+    enable_py_buffer_protocol();
+    PyObjectPtr<PyObject> const num_read_byte_obj{PyObject_CallMethod(
+            m_input_ir_stream,
+            "readinto",
+            "O",
+            py_reinterpret_cast<PyObject>(this)
+    )};
+    disable_py_buffer_protocol();
+
+    if (nullptr == num_read_byte_obj.get()) {
+        return false;
+    }
+    num_bytes_read = PyLong_AsSsize_t(num_read_byte_obj.get());
+    if (0 > num_bytes_read) {
+        return false;
+    }
+    m_buffer_size += num_bytes_read;
+    return true;
+}
+
+auto PyDecoderBuffer::py_getbuffer(Py_buffer* view, int flags) -> int {
+    // Don't need to set the exception message during the failure.
+    // The Python level caller will set the exception and thus overwrite it.
+    if (false == is_py_buffer_protocol_enabled()) {
+        return -1;
+    }
+    auto const buffer_size{m_buffer_capacity - m_buffer_size};
+    if (0 >= buffer_size) {
+        return -1;
+    }
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+    auto* buffer{m_read_buffer + m_buffer_size};
+    return PyBuffer_FillInfo(
+            view,
+            py_reinterpret_cast<PyObject>(this),
+            buffer,
+            buffer_size,
+            0,
+            flags
+    );
+}
+
+auto PyDecoderBuffer::commit_read_buffer_consumption(Py_ssize_t num_bytes_consumed) -> bool {
+    if (get_num_unconsumed_bytes() < num_bytes_consumed) {
+        PyErr_SetString(PyExc_OverflowError, cDecoderBufferOverflowError);
+        return false;
+    }
+    m_num_current_bytes_consumed += num_bytes_consumed;
+    return true;
+}
+
+auto PyDecoderBuffer::test_streaming(unsigned seed) -> PyObject* {
+    std::default_random_engine rand_generator(seed);
+    std::vector<uint8_t> read_bytes;
+    bool reach_istream_end{false};
+    while (false == reach_istream_end) {
+        std::uniform_int_distribution<Py_ssize_t> distribution(1, m_buffer_size);
+        auto num_bytes_to_read{distribution(rand_generator)};
+        if (get_num_unconsumed_bytes() < num_bytes_to_read) {
+            Py_ssize_t num_bytes_read_from_istream{0};
+            if (false == populate_read_buffer(num_bytes_read_from_istream)) {
+                return nullptr;
+            }
+            if (0 == num_bytes_read_from_istream) {
+                reach_istream_end = true;
+            }
+            num_bytes_to_read = std::min<Py_ssize_t>(num_bytes_to_read, m_buffer_size);
+        }
+        auto* unconsumed_bytes{get_unconsumed_bytes()};
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        read_bytes.insert(read_bytes.end(), unconsumed_bytes, unconsumed_bytes + num_bytes_to_read);
```

Comment:
This appears to only rely on new code and isn't directly limited by cpython. Why are we disabling the linting for it?

Reply:
Resolved by using `gsl::span`